File: Spark/Ejercicios/chapter3/ejemplo3.scala
package main.scala.chapter3.ejemplo3

import org.apache.spark.sql.types._
import org.apache.spark.sql.SparkSession

// Leer datos de un JSON

object ejemplo {
    def main(args: Array[String]) {
        val spark = SparkSession
            .builder
            .appName("Ejemplo3")
            .getOrCreate()

        if (args.length <= 0) {
            println("uso: Ejemplo3 <ruta a blogs.json>")
            System.exit(1)
        }

        // Path al JSON
        val jsonFile = args(0)

        // Definición programática del schema
        val schema = StructType(Array(StructField("Id",        IntegerType, false),
                                      StructField("First",     StringType,  false),
                                      StructField("Last",      StringType,  false),
                                      StructField("Url",       StringType,  false),
                                      StructField("Published", StringType,  false),
                                      StructField("Hits",      IntegerType, false),
                                      StructField("Campaigns", ArrayType(StringType), false)))

        // Creamos un DataFrame leyendo del JSON
        val blogsDF = spark.read.schema(schema).json(jsonFile)

        blogsDF.show(false)

        println(blogsDF.printSchema)
        println(blogsDF.schema)

    }
}

File: Spark/Ejercicios/chapter3/ejemplo3.py
import sys

from pyspark.sql.types import *
from pyspark.sql import SparkSession

if __name__ == "__main__":

    spark = (SparkSession.builder.appName("Ejemplo3").getOrCreate())

    if len(sys.argv) != 2:
        print("uso: Ejemplo3 <ruta a blogs.json>", file=sys.stderr)
        sys.exit(-1)

    jsonFile = sys.argv[1]
    print(jsonFile)

    schema = StructType([StructField("Id", IntegerType(), False),
                         StructField("First", StringType(), False),
                         StructField("Last", StringType(), False),
                         StructField("Url", StringType(), False),
                         StructField("Published", StringType(), False),
                         StructField("Hits", IntegerType(), False),
                         StructField("Campaigns", ArrayType(StringType()), False)
                         ])

    blogsDF = spark.read.schema(schema).json(jsonFile)

    blogsDF.show()

    blogsDF.printSchema()
    print(blogsDF.schema)